streams/src/test/java/org/apache/kafka/streams/integration/CustomHandlerIntegrationTest.java

```diff
@@ -0,0 +1,167 @@
+
+    @BeforeAll
+    public static void startCluster() throws IOException {
+        CLUSTER.start();
+    }
+
+    @AfterAll
+    public static void closeCluster() {
+        CLUSTER.stop();
+    }
+
+    private final long timeout = 60000;
+
+    // topic name
+    private static final String STREAM_INPUT = "STREAM_INPUT";
+    private static final String NON_EXISTING_TOPIC = "non_existing_topic";
+
+    private KafkaStreams kafkaStreams;
+    AtomicReference<Throwable> caughtException;
+    Topology topology;
+    private String appId;
+
+    @BeforeEach
+    public void before(final TestInfo testInfo) throws InterruptedException {
+        final StreamsBuilder builder = new StreamsBuilder();
+        CLUSTER.createTopics(STREAM_INPUT);
+        caughtException = new AtomicReference<>();
+        final String safeTestName = safeUniqueTestName(testInfo);
+        appId = "app-" + safeTestName;
+
+
+        builder.stream(STREAM_INPUT, Consumed.with(Serdes.Integer(), Serdes.String()))
+                .to(NON_EXISTING_TOPIC, Produced.with(Serdes.Integer(), Serdes.String()));
+        produceRecords();
+        topology = builder.build();
+    }
+
+    @AfterEach
+    public void after() throws InterruptedException {
+        CLUSTER.deleteTopics(STREAM_INPUT);
+        if (kafkaStreams != null) {
+            kafkaStreams.close();
+            kafkaStreams.cleanUp();
+        }
+    }
+
+    private void produceRecords() {
+        final Properties props = TestUtils.producerConfig(
+                CLUSTER.bootstrapServers(),
+                IntegerSerializer.class,
+                StringSerializer.class,
+                new Properties());
+        IntegrationTestUtils.produceKeyValuesSynchronouslyWithTimestamp(
+                STREAM_INPUT,
+                Collections.singletonList(new KeyValue<>(1, "A")),
+                props,
+                CLUSTER.time.milliseconds() + 2
+        );
+    }
+    private Properties getCommonProperties() {
+        final Properties streamsConfiguration = new Properties();
+        streamsConfiguration.put(StreamsConfig.APPLICATION_ID_CONFIG, appId);
+        streamsConfiguration.put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, CLUSTER.bootstrapServers());
+        streamsConfiguration.put(StreamsConfig.DEFAULT_KEY_SERDE_CLASS_CONFIG, Serdes.Integer().getClass());
+        streamsConfiguration.put(StreamsConfig.DEFAULT_VALUE_SERDE_CLASS_CONFIG, Serdes.String().getClass());
+        streamsConfiguration.put(StreamsConfig.NUM_STREAM_THREADS_CONFIG, NUM_THREADS);
+        return streamsConfiguration;
+    }
+
+    private void closeApplication(final Properties streamsConfiguration) throws Exception {
+        kafkaStreams.close();
+        kafkaStreams.cleanUp();
+        IntegrationTestUtils.purgeLocalStreamsState(streamsConfiguration);
+        final long timeout = 60000;
+        TestUtils.waitForCondition(
+                () -> kafkaStreams.state() == State.NOT_RUNNING,
```

Comment:
Given that the uncaught exception handler return `SHUTDOWN_CLIENT` the final state should be `ERROR` IIRC?

Reply:
exactly